compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs

```diff
@@ -1148,6 +1149,37 @@ pub fn typeid_for_instance<'tcx>(
                 tcx.mk_args_trait(invoke_ty, trait_ref.args.into_iter().skip(1));
             instance.args = instance.args.rebase_onto(tcx, impl_id, abstract_trait_args);
         }
+    } else if tcx.is_closure_like(instance.def_id()) {
+        // We're either a closure or a coroutine. Our goal is to find the trait we're defined on,
+        // instantiate it, and take the type of its only method as our own.
+        let closure_ty = instance.ty(tcx, ty::ParamEnv::reveal_all());
+        let (trait_id, tuple_args) = match closure_ty.kind() {
+            ty::Closure(_def_id, args) => {
+                let closure_args = ty::ClosureArgs { args };
+                let trait_id = tcx.fn_trait_kind_to_def_id(closure_args.kind()).unwrap();
+                let tuple_args = closure_args.sig().inputs().no_bound_vars().unwrap()[0];
+                (trait_id, tuple_args)
+            }
+            ty::Coroutine(_def_id, args) => (
+                tcx.require_lang_item(LangItem::Coroutine, None),
+                ty::CoroutineArgs { args }.resume_ty(),
+            ),
```

Comment:
You should be able to support `CoroutineClosure`, and use `async_fn_trait_kind_to_def_id` (wait, you'll need to add that).

Reply:
Looking into that, is that correct given that `AsyncFn` is not object-safe? Presumably my `trait_object_ty` will fail to generate an appropriate `dyn` type because the returned future is a GAT?

Reply:
Ah, didn't see this comment. Yeah, huh, I guess that makes sense; they are in fact not object safe, though you can call them through `FnOnce`, and I think we need to support that. Not exactly sure, may need playing around; though you could defer that 'til later I guess.

Reply:
I added a test for calling through `FnOnce`.